### PR TITLE
Enable to run TPCDS performance benchmarks on GCP

### PR DESCRIPTION
**Pull request by grzegorz8**
## Description

This PR enables to run TCPDS performance benchmarks not only on AWS but also on Google Cloud Platform.This is an extension of the framework introduced by: https://github.com/delta-io/delta/pull/973 .

In order to run the benchmark, you need to manually create a Dataproc cluster first. All prerequisites and sample gcloud commands are in README file. After that you can run *Load data* and *Query data* steps provided by the framework.

Please see the README updates for more details.

## How was this patch tested?

I manually ran the benchmarks on GCP.

## Does this PR introduce _any_ user-facing changes?

No.

### CODE REVIEW

File: benchmarks/build.sbt [hunk 1]
```diff
@@ -20,7 +20,7 @@ scalaVersion := "2.12.14"
 lazy val root = (project in file("."))
   .settings(
     name := "benchmarks",
-    libraryDependencies += "org.apache.spark" %% "spark-sql" % "3.2.0" % "provided",
```
Comment (dennyglee) [inline benchmarks/build.sbt:23]: QQ - why switch to Spark 3.1.2 vs. Spark 3.2.0?   Was this an issue related to Dataproc?

Comment (dennyglee) [reply benchmarks/build.sbt:23]: Got it - makes sense - thanks!

**Review by dennyglee — APPROVED** [review]
LGTM!

File: benchmarks/run-benchmark.py [hunk 2]
```diff
@@ -79,12 +87,20 @@
         help="Local path to delta repository which will be used for running the benchmark " +
              "instead of the version specified in the specification. Make sure that new delta" +
              " version is compatible with version in the spec.")
+    parser.add_argument(
+        "--cloud-provider",
+        choices=delta_log_store_classes.keys(),
+        help="Cloud where the benchmark will be executed.")
```
Comment (tdas) [inline benchmarks/run-benchmark.py:93]: since there is no default here, users must provide this right? have you made sure that the examples in the readme always provides this?

Comment (grzegorz8) [reply benchmarks/run-benchmark.py:93]: Good catch. I'll update usage examples as well. Thanks!

File: benchmarks/README.md [hunk 3]
```diff
@@ -44,8 +46,86 @@ Once the EMR cluster is ready, note the following:
...
+    --worker-boot-disk-type pd-ssd \
+    --worker-boot-disk-size 100 \
+    --num-worker-local-ssds 4 \
+    --worker-local-ssd-interface NVME \
+    --dataproc-metastore projects/<project-name>/locations/<region>/services/dataproc-metastore-for-benchmarks \
+    --enable-component-gateway \
+    --image-version 2.0-debian10
+```
+
+#### Input data
```
Comment (grzegorz8) [inline benchmarks/README.md:114]: Now I thinking about having a permanent copy of TPC-DS data in Databricks' Google Storage bucket, marked as "Requester Pays", until Google Dataproc supports Hadoop >= 3.3.4, instead of asking users to copy the data to their buckets first. WDYT @dennyglee ?

Comment (dennyglee) [reply benchmarks/README.md:114]: This is a good idea but let's create a new issue to put the data into an open GCS bucket

File: benchmarks/run-benchmark.py (hunk 2 above)
Comment (grzegorz8) [reply benchmarks/run-benchmark.py:93]: Updated.

Comment (grzegorz8) [reply benchmarks/run-benchmark.py:93]: @tdas Could you re-review the PR, please? Thanks!

File: benchmarks/README.md [hunk 4]
```diff
@@ -1,25 +1,27 @@
...
-1. *Load data*: You have to create the TPC-DS database with all the Delta tables. To do that, the raw TPC-DS data has been provided as Apache Parquet files. In this step you will have to use your EMR cluster to read the parquet files and rewrite them as Delta tables.
+1. *Load data*: You have to create the TPC-DS database with all the Delta tables. To do that, the raw TPC-DS data has been provided as Apache Parquet files. In this step you will have to use your EMR or a Dataproc cluster to read the parquet files and rewrite them as Delta tables.
 2. *Query data*: Then, using the tables definitions in the Hive Metatore, you can run the 99 benchmark queries.   
 
-The next section will provide the detailed steps of how to setup the necessary Hive Metastore and EMR cluster, how to test the setup with small-scale data, and then finally run the full scale benchmark.  
+The next section will provide the detailed steps of how to setup the necessary Hive Metastore and a cluster, how to test the setup with small-scale data, and then finally run the full scale benchmark.
 
-### Prerequisites
+### Amazon Web Services
```
Comment (tdas) [inline benchmarks/README.md:14]: I think it should clarify and say "Configure cluster with Amazon Web Services"... so that its consistent with the next item at this level `Test the cluster setup

File: benchmarks/README.md [hunk 5]
```diff
@@ -44,99 +46,237 @@ Once the EMR cluster is ready, note the following:
...
 These will be needed to run the workloads in this framework. 
 
-### Prepare S3 bucket
-Create a new S3 bucket (or use an existing one) which is in the same region your EMR cluster.
+#### Prepare S3 bucket
+Create a new S3 bucket (or use an existing one) which is in the same region as your EMR cluster.
+
+_________________
+
+### Google Cloud Platform
```
Comment (tdas) [inline benchmarks/README.md:54]: similar change as suggested above

File: benchmarks/README.md [hunk 6]
```diff
@@ -44,99 +46,237 @@ Once the EMR cluster is ready, note the following:
...
+
+#### Prerequisites
+- A GCP account with necessary permissions to do the following:
+  - Manage Dataproc clusters for running the benchmark
+  - Manage Dataproc Metastore instances
+  - Read and write to a GCS bucket from the Dataproc cluster
+- A GCS bucket which will be used to generate the TPC-DS data.
+- A machine which has access to the GCP setup and where this repository has been downloaded or cloned.
+- SSH keys for a user which will be used to access the master node. The user's SSH key can be either [a project-wide key](https://cloud.google.com/compute/docs/connect/add-ssh-keys#add_ssh_keys_to_project_metadata)
```
Comment (tdas) [inline benchmarks/README.md:63]: is this the PEM file?

File: benchmarks/run-benchmark.py [hunk 7]
```diff
@@ -79,12 +87,20 @@
...
+    parser.add_argument(
+        "--ssh-user",
```
Comment (tdas) [inline benchmarks/run-benchmark.py:95]: why was this needed? i dont see this being used in the instructions

File: benchmarks/scripts/benchmarks.py [hunk 8]
```diff
@@ -225,52 +259,76 @@ def upload_jar_to_cluster(self, cluster_hostname, ssh_id_file, delta_version_to_
...
         run_cmd(scp_cmd, stream_output=True)
         print(">>> Benchmark JAR uploaded to cluster\n")
         return f"~/{jar_remote_path}"
 
-    def start_benchmark_via_ssh(self, cluster_hostname, ssh_id_file, jar_path):
+    def install_dependencies_via_ssh(self, cluster_hostname, ssh_id_file, ssh_user):
+        script_file_name = f"{self.benchmark_id}-install-deps.sh"
+        script_file_text = """
+#!/bin/bash
+packagesNeeded='screen'
```
Comment (tdas) [inline benchmarks/scripts/benchmarks.py:272]: nit: this script will attempt to install screen everytime the benchmark is run. it might be better to check whether screen exists first before attempting to install it.

File: benchmarks/scripts/benchmarks.py [hunk 9]
```diff
@@ -225,52 +259,76 @@ def upload_jar_to_cluster(self, cluster_hostname, ssh_id_file, delta_version_to_
...
+            script_file.write(script_file_text)
+            script_file.close()
+
+            scp_cmd = (
+                    f"scp -i {ssh_id_file} {script_file_name}" +
+                    f" {ssh_user}@{cluster_hostname}:{script_file_name}"
+            )
+            print(scp_cmd)
+            run_cmd(scp_cmd, stream_output=True)
+            run_cmd(f"ssh -i {ssh_id_file} {ssh_user}@{cluster_hostname} chmod +x {script_file_name}")
```
Comment (tdas) [inline benchmarks/scripts/benchmarks.py:327]: I think there is a util method run_cmd_over_ssh... can that be used here?

**Review by tdas — APPROVED** [review]
approved, but a few minor comments that wouuld be good to resolve before we merge this

File: benchmarks/scripts/benchmarks.py (hunk 8 above)
Comment (grzegorz8) [reply benchmarks/scripts/benchmarks.py]: I extended the bash script as suggested.
